## IMPLEMENTATION OF IMAGE CLASSIFIER OF DATASET : CIFAR-10 USING PYTORCH

##### TRY 1: USING NEURAL NETWORK 

In [3]:
# Importing necessary libraries
# This includes the essential PyTorch and torchvision modules for working with neural networks and datasets.
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
import torchvision
import torchvision.transforms as transforms

# Define the transform
transform = transforms.Compose([
    transforms.ToTensor(),  # Convert images to PyTorch tensors
    transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))  # Normalize image pixel values
])

# Download and load the CIFAR-10 dataset
trainset = torchvision.datasets.CIFAR10(root='./data', train=True, download=True, transform=transform) # Create training dataset
trainloader = torch.utils.data.DataLoader(trainset, batch_size=4, shuffle=True, num_workers=2)

testset = torchvision.datasets.CIFAR10(root='./data', train=False, download=True, transform=transform) # Create test dataset
testloader = torch.utils.data.DataLoader(testset, batch_size=4, shuffle=False, num_workers=2)

# Define the classes for classification
classes = ('plane', 'car', 'bird', 'cat', 'deer', 'dog', 'frog', 'horse', 'ship', 'truck')

# Define the neural network architecture
class Net(nn.Module):
    def __init__(self):
        # Call the constructor of the parent class (nn.Module) to initialize the base class
        super(Net, self).__init__()

        # Convolutional layers
        self.conv1 = nn.Conv2d(3, 6, 5)  # Input channels: 3, Output channels: 6, Kernel size: 5x5
        self.pool = nn.MaxPool2d(2, 2)  # Max pooling layer with kernel size 2x2 and stride 2
        self.conv2 = nn.Conv2d(6, 16, 5)  # Input channels: 6, Output channels: 16, Kernel size: 5x5

        # Fully connected layers
        # Flatten layer to convert 3D tensor to 1D tensor before fully connected layers
        self.fc1 = nn.Linear(16 * 5 * 5, 120)  # Input features: 16*5*5, Output features: 120
        self.fc2 = nn.Linear(120, 84)  # Input features: 120, Output features: 84
        self.fc3 = nn.Linear(84, 10)  # Input features: 84, Output features: 10 (number of classes)

    def forward(self, x):
        # Forward pass through the network

        # First convolutional layer (conv1) followed by ReLU activation and max pooling
        x = self.pool(F.relu(self.conv1(x)))
        # Second convolutional layer (conv2) followed by ReLU activation and max pooling
        x = self.pool(F.relu(self.conv2(x)))
        # Flatten the output to a 1D tensor before passing it to fully connected layers
        x = x.view(-1, 16 * 5 * 5)
        # First fully connected layer (fc1) followed by ReLU activation
        x = F.relu(self.fc1(x))
        # Second fully connected layer (fc2) followed by ReLU activation
        x = F.relu(self.fc2(x))
        # Final fully connected layer (fc3) for classification output
        x = self.fc3(x)

        return x # Return the final output 
    
# Instantiate the network
net = Net()

# Define the loss function and optimizer
criterion = nn.CrossEntropyLoss()  # Loss function for multi-class classification
optimizer = optim.SGD(net.parameters(), lr=0.001, momentum=0.9)  # SGD optimizer with momentum
epochs = 10  # Number of times to iterate through the entire dataset during training

# Training loop
for epoch in range(epochs):  # Loop over the dataset for a specified number of epochs
    running_loss = 0.0  # Initialize the running loss for the current epoch
    for i, data in enumerate(trainloader, 0):  # Iterate over batches in the training loader
        inputs, labels = data  # Get inputs and labels for the current batch

        optimizer.zero_grad()  # Zero the gradients to clear previous gradients

        outputs = net(inputs)  # Forward pass to compute the predicted outputs
        loss = criterion(outputs, labels)  # Compute the loss between predicted and true labels
        loss.backward()  # Backward pass to compute gradients of the loss with respect to model parameters
        optimizer.step()  # Update model parameters using the optimizer

        running_loss += loss.item()  # Accumulate the running loss for statistics
        if i % 2000 == 1999:  # Print every 2000 batches
            print('[%d, %5d] loss: %.3f' % (epoch + 1, i + 1, running_loss / 2000))
            running_loss = 0.0  # Reset the running loss for the next set of batches

# Testing the model
correct = 0  # Initialize the number of correctly predicted samples
total = 0  # Initialize the total number of samples

# Use torch.no_grad() to disable gradient computation during testing
with torch.no_grad():
    for data in testloader:  # Iterate over batches in the test loader
        images, labels = data  # Get inputs and true labels for the current batch
        outputs = net(images)  # Forward pass to compute predicted outputs
        _, predicted = torch.max(outputs.data, 1)  # Get the index of the maximum predicted value
        total += labels.size(0)  # Increment the total number of samples by the batch size
        correct += (predicted == labels).sum().item()  # Count the number of correctly predicted samples

accuracy = 100 * correct / total # Calculate accuracy
print(f"Accuracy on the test set: {accuracy:.2f}%") # Print the accuracy on the test set

Files already downloaded and verified
Files already downloaded and verified
[1,  2000] loss: 2.227
[1,  4000] loss: 1.848
[1,  6000] loss: 1.671
[1,  8000] loss: 1.573
[1, 10000] loss: 1.519
[1, 12000] loss: 1.462
[2,  2000] loss: 1.420
[2,  4000] loss: 1.357
[2,  6000] loss: 1.330
[2,  8000] loss: 1.336
[2, 10000] loss: 1.293
[2, 12000] loss: 1.316
[3,  2000] loss: 1.235
[3,  4000] loss: 1.209
[3,  6000] loss: 1.211
[3,  8000] loss: 1.203
[3, 10000] loss: 1.185
[3, 12000] loss: 1.188
[4,  2000] loss: 1.114
[4,  4000] loss: 1.118
[4,  6000] loss: 1.126
[4,  8000] loss: 1.126
[4, 10000] loss: 1.118
[4, 12000] loss: 1.103
[5,  2000] loss: 1.033
[5,  4000] loss: 1.051
[5,  6000] loss: 1.051
[5,  8000] loss: 1.056
[5, 10000] loss: 1.073
[5, 12000] loss: 1.046
[6,  2000] loss: 0.977
[6,  4000] loss: 0.988
[6,  6000] loss: 0.979
[6,  8000] loss: 1.000
[6, 10000] loss: 1.018
[6, 12000] loss: 1.013
[7,  2000] loss: 0.917
[7,  4000] loss: 0.944
[7,  6000] loss: 0.941
[7,  8000] loss: 0.967
[7, 

#### TRY 2: USING CNN NEURAL NETWORK AND DOING MODIFACTIONS FOR BETTER ACCURACY 

In [22]:
# Import necessary libraries
# This includes the essential PyTorch and torchvision modules for working with neural networks and datasets.
import torch
import torch.nn as nn
import torch.optim as optim
import torchvision.transforms as transforms
import torchvision.datasets as datasets
from torch.utils.data import DataLoader

# Define the CNN architecture
class SimpleCNN(nn.Module):
    def __init__(self):
        super(SimpleCNN, self).__init__()
    
        # Convolutional layer 1: Input channels=3, output channels=32, kernel size=3, padding=1
        self.conv1 = nn.Conv2d(3, 32, kernel_size=3, padding=1)
        # Rectified Linear Unit (ReLU) activation function
        self.relu = nn.ReLU()
        # Max pooling layer 1: Kernel size=2, stride=2
        self.maxpool = nn.MaxPool2d(2)
        # Convolutional layer 2: Input channels=32, output channels=64, kernel size=3, padding=1
        self.conv2 = nn.Conv2d(32, 64, kernel_size=3, padding=1)
        # Flatten layer to convert 3D tensor to 1D tensor
        self.flatten = nn.Flatten()
        # Fully connected layer 1: Input features=64*8*8, output features=512
        self.fc1 = nn.Linear(64 * 8 * 8, 512)
        # Fully connected layer 2: Input features=512, output features=10 (output classes)
        self.fc2 = nn.Linear(512, 10)

    def forward(self, x):

        # Convolutional layer 1: Input tensor x undergoes convolution
        x = self.conv1(x)
        # Apply ReLU activation function to introduce non-linearity
        x = self.relu(x)
        # Perform max pooling to down-sample the spatial dimensions
        x = self.maxpool(x)
    
        # Convolutional layer 2: Apply another convolution operation
        x = self.conv2(x)
        # Apply ReLU activation
        x = self.relu(x)
        # Another max pooling operation
        x = self.maxpool(x)
    
        # Flatten the tensor to prepare for fully connected layers
        x = self.flatten(x)
    
        # Fully connected layer 1: Apply linear transformation
        x = self.fc1(x)
        # Apply ReLU activation
        x = self.relu(x)
    
        # Fully connected layer 2: Produce the final output
        x = self.fc2(x)
    
        # Return the final output tensor after passing through the network
        return x

# Set device (GPU if available, otherwise CPU)
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
# Hyperparameters
batch_size = 64 # Set the batch size for training
learning_rate = 0.001 # Set the learning rate for the optimizer
epochs = 10 # Set the number of training epochs

# Define data transformation for CIFAR-10 dataset
transform = transforms.Compose([
    transforms.ToTensor(),  # Convert images to PyTorch tensors
    transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))  # Normalize pixel values
])

# Download and Load CIFAR-10 training and test datasets
train_dataset = datasets.CIFAR10(root='./data', train=True, download=True, transform=transform)  # Create training dataset
test_dataset = datasets.CIFAR10(root='./data', train=False, download=True, transform=transform)  # Create test dataset

# Create DataLoader instances for training and test datasets
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

# Initialize the neural network model, loss function, and optimizer
model = SimpleCNN().to(device)  # Move the model to GPU if available
criterion = nn.CrossEntropyLoss()  # Cross-entropy loss for classification
optimizer = optim.Adam(model.parameters(), lr=learning_rate)  # Adam optimizer

# Train the model
for epoch in range(epochs): # Iterate over each epoch
    model.train()  # Set the model to training mode
    for images, labels in train_loader: # Iterate through batches of training data
        images, labels = images.to(device), labels.to(device) # Move data to GPU if available
        optimizer.zero_grad()  # Zero gradients to clear previous gradients
        outputs = model(images)  # Forward pass
        loss = criterion(outputs, labels)  # Compute the loss
        loss.backward()  # Backward pass to compute gradients
        optimizer.step()  # Update model parameters using the optimizer

    print(f"Epoch {epoch+1}/{epochs}, Loss: {loss.item()}")  # Print the loss for the current epoch

# Evaluate the model on the test set
model.eval() # Set the model to evaluation mode
correct = 0  # Initialize the number of correctly predicted samples
total = 0  # Initialize the total number of samples

with torch.no_grad(): # Use torch.no_grad() to disable gradient computation during testing
    for images, labels in test_loader: # Iterate through batches of test data
        images, labels = images.to(device), labels.to(device) # Move data to GPU if available
        outputs = model(images) # Forward pass through the model
        _, predicted = torch.max(outputs.data, 1) # Get the predicted class labels
        total += labels.size(0) # Increment total number of images
        correct += (predicted == labels).sum().item() # Count correctly predicted images

accuracy = correct / total # Calculate accuracy
print(f"Test Accuracy: {accuracy * 100:.2f}%") # Print the accuracy on the test set


Files already downloaded and verified
Files already downloaded and verified
Epoch 1/10, Loss: 0.9492315053939819
Epoch 2/10, Loss: 1.301919937133789
Epoch 3/10, Loss: 0.4410093128681183
Epoch 4/10, Loss: 0.4600154757499695
Epoch 5/10, Loss: 0.3084782361984253
Epoch 6/10, Loss: 0.38129258155822754
Epoch 7/10, Loss: 0.21933627128601074
Epoch 8/10, Loss: 0.09921213984489441
Epoch 9/10, Loss: 0.006839558482170105
Epoch 10/10, Loss: 0.14239855110645294
Test Accuracy: 72.15%


#### TRY 3: USING PRETRAINED MODEL: RESNET 18 FOR ACHIEVING BETTER ACCURACY

In [19]:
# Importing necessary libraries
# This includes the essential PyTorch and torchvision modules for working with neural networks and datasets.
import torch
import torch.nn as nn
import torch.optim as optim
import torchvision
import torchvision.transforms as transforms
import torchvision.models as models

# Define the transform for data augmentation and normalization
transform = transforms.Compose([
    transforms.RandomHorizontalFlip(),  # Randomly flip the image horizontally
    transforms.RandomResizedCrop(32),    # Randomly crop the image and resize to 32x32
    transforms.ToTensor(),               # Convert image to PyTorch tensor
    transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))  # Normalize image pixel values
])

# Download and load the CIFAR-10 dataset
trainset = torchvision.datasets.CIFAR10(root='./data', train=True, download=True, transform=transform) # Create training dataset
trainloader = torch.utils.data.DataLoader(trainset, batch_size=64, shuffle=True, num_workers=2)

testset = torchvision.datasets.CIFAR10(root='./data', train=False, download=True, transform=transform) # Create test dataset
testloader = torch.utils.data.DataLoader(testset, batch_size=64, shuffle=False, num_workers=2)

# Use a pre-trained ResNet18 model
model = models.resnet18(pretrained=True)
num_ftrs = model.fc.in_features
model.fc = nn.Linear(num_ftrs, 10)  # Change the output layer to have 10 classes

# Move the model to GPU if available
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu") # Check if GPU is available, else use CPU
model = model.to(device)

# Define the loss function and optimizer
criterion = nn.CrossEntropyLoss()  # Loss function for multi-class classification
optimizer = optim.SGD(model.parameters(), lr=0.001, momentum=0.9)  # SGD optimizer with momentum

# Training loop
num_epochs = 10  # Set the number of training epochs
# Iterate over each epoch
for epoch in range(num_epochs):
    model.train()  # Set the model to training mode
    running_loss = 0.0  # Variable to track the running loss during each epoch
    # Iterate over the training dataset
    for i, data in enumerate(trainloader, 0): 
        inputs, labels = data # Get inputs and labels for the current batch
        inputs, labels = inputs.to(device), labels.to(device) # Move data to GPU if available
        optimizer.zero_grad()  # Clear previous gradients
        outputs = model(inputs)  # Forward pass
        loss = criterion(outputs, labels)  # Calculate the loss
        loss.backward()  # Backward pass to compute gradients
        optimizer.step()  # Update model parameters using the optimizer
        running_loss += loss.item()  # Accumulate the running loss

    # Print the average loss for the current epoch
    print(f"Epoch {epoch + 1}, Loss: {running_loss / len(trainloader)}")


# Testing the model
model.eval()  # Set the model to evaluation mode
correct = 0  # Variable to track the number of correctly predicted images
total = 0  # Variable to track the total number of images in the test set

with torch.no_grad(): # Use torch.no_grad() to disable gradient computation during testing
    for data in testloader:  # Iterate over batches in the test loader
        images, labels = data # Get inputs and labels for the current batch
        images, labels = images.to(device), labels.to(device)  # Move data to GPU if available
        outputs = model(images)  # Forward pass
        _, predicted = torch.max(outputs.data, 1)  # Get predicted class labels
        total += labels.size(0)  # Increment total number of images
        correct += (predicted == labels).sum().item()  # Count correctly predicted images

accuracy = 100 * correct / total # Calculate accuracy
print(f"Accuracy on the test set: {accuracy:.2f}%") # Print the test accuracy


Files already downloaded and verified
Files already downloaded and verified
Epoch 1, Loss: 1.3403009552784892
Epoch 2, Loss: 1.0638465591708717
Epoch 3, Loss: 0.9708040430570197
Epoch 4, Loss: 0.9199596481859836
Epoch 5, Loss: 0.8820667493983608
Epoch 6, Loss: 0.8590083898943098
Epoch 7, Loss: 0.8284356531203555
Epoch 8, Loss: 0.8084617858500127
Epoch 9, Loss: 0.788709190228711
Epoch 10, Loss: 0.7772868377016023
Accuracy on the test set: 72.46%


#### TRY 4: USING ANOTHER PRETRAINED MODEL: DENSENET 121 MODEL FOR BETTER ACCURACY

In [2]:
# Importing necessary libraries 
import torch
import torch.nn as nn
import torch.optim as optim
import torchvision.transforms as transforms
import torchvision.datasets as datasets
from torch.utils.data import DataLoader
from torchvision import models

# Step 1: Set device and hyperparameters
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")  # Check if GPU is available, else use CPU
batch_size = 64  # Number of images in each mini-batch
learning_rate = 0.001  # Learning rate for the optimizer
epochs = 10  # Number of times to iterate through the entire dataset during training

# Step 2: Load and preprocess the CIFAR-10 dataset
transform = transforms.Compose([
    transforms.ToTensor(),  # Convert images to PyTorch tensors
    transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))  # Normalize image pixel values
])

# Download and create training and test datasets
train_dataset = datasets.CIFAR10(root='./data', train=True, download=True, transform=transform) # Create training dataset
test_dataset = datasets.CIFAR10(root='./data', train=False, download=True, transform=transform) # Create test dataset

# Create DataLoader instances to efficiently load and iterate over batches of data
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False)

# Step 3: Initialize the pre-trained DenseNet model
model = models.densenet121(pretrained=True)  # Load pre-trained DenseNet model
# Modify the classifier for CIFAR-10 (10 classes)
model.classifier = nn.Linear(1024, 10)  # Change the output layer to have 10 classes for CIFAR-10
model = model.to(device)  # Move the model to the GPU (if available)

# Step 4: Set up loss function and optimizer
criterion = nn.CrossEntropyLoss()  # Loss function for multi-class classification
optimizer = optim.Adam(model.parameters(), lr=learning_rate)  # Adam optimizer with specified learning rate

# Step 5: Train the model
for epoch in range(epochs): # Loop over the dataset for a specified number of epochs
    model.train()  # Set the model to training mode
    for images, labels in train_loader: # Iterate through batches of training data
        images, labels = images.to(device), labels.to(device)  # Move data to GPU if available

        optimizer.zero_grad()  # Zero gradients to clear previous gradients
        outputs = model(images)  # Forward pass
        loss = criterion(outputs, labels)  # Compute the loss
        loss.backward()  # Backward pass to compute gradients
        optimizer.step()  # Update model parameters using the optimizer

    print(f"Epoch {epoch + 1}/{epochs}, Loss: {loss.item()}")  # Print loss after each epoch

# Step 6: Evaluate the model on the test set
model.eval()  # Set the model to evaluation mode
correct = 0  # Initialize the number of correctly predicted samples
total = 0  # Initialize the total number of samples

# Use torch.no_grad() to disable gradient computation during testing
with torch.no_grad():
    for images, labels in test_loader: # Iterate over batches in the test loader
        images, labels = images.to(device), labels.to(device)  # Move data to GPU if available
        outputs = model(images)  # Forward pass
        _, predicted = torch.max(outputs.data, 1)  # Get predicted class labels
        total += labels.size(0)  # Increment total number of images
        correct += (predicted == labels).sum().item()  # Count correctly predicted images

accuracy = correct / total  # Calculate accuracy
print(f"Test Accuracy: {accuracy * 100:.2f}%")  # Print test accuracy


Files already downloaded and verified
Files already downloaded and verified


c:\Users\divig\anaconda3\Lib\site-packages\torchvision\models\_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
c:\Users\divig\anaconda3\Lib\site-packages\torchvision\models\_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=DenseNet121_Weights.IMAGENET1K_V1`. You can also use `weights=DenseNet121_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)
Downloading: "https://download.pytorch.org/models/densenet121-a639ec97.pth" to C:\Users\divig/.cache\torch\hub\checkpoints\densenet121-a639ec97.pth
100%|██████████| 30.8M/30.8M [00:08<00:00, 3.61MB/s]


Epoch 1/10, Loss: 0.7561923265457153
Epoch 2/10, Loss: 0.17223471403121948
Epoch 3/10, Loss: 0.3745187222957611
Epoch 4/10, Loss: 0.5094252824783325
Epoch 5/10, Loss: 0.16775503754615784
Epoch 6/10, Loss: 0.2698609232902527
Epoch 7/10, Loss: 0.1625276356935501
Epoch 8/10, Loss: 0.5435947179794312
Epoch 9/10, Loss: 0.18278992176055908
Epoch 10/10, Loss: 0.5831239223480225
Test Accuracy: 84.48%


### IN CONCLUSION,I FOUND OUT THAT THROUGH THE PRETRAINED MODEL DENSENET121 I ACHIEVED THE BEST ACCURACY AMONG ALL THE OTHER WAYS WHICH IS EQUAL TO 84.48% ACCURACY WHICH IS QUITE IMPRESSIVE 